File: 2020/ruby/01/main.rb
require 'pathname'

class ExpenseReporter

  attr_accessor :expenses, :expense1, :expense2, :expense3

  DESIRED_EXPENSE = 2020;

  # @input [Pathname] filename
  def initialize(filename)
    @expenses = filename.read.split("\n")
  end

  def two_expenses
    reset_expenses

    expenses.each_with_index do |exp1, idx|
      @expense1 = exp1.to_i
      expenses.each_with_index do |exp2, idx2|
        @expense2 = exp2.to_i
        break if (expense1 + expense2) == DESIRED_EXPENSE
      end

      break if (expense1 + expense2) == DESIRED_EXPENSE
    end

    expense1 * expense2
  end

  # TODO: Refactor
  def three_expenses
    reset_expenses
    expenses.each_with_index do |exp1, idx|
      @expense1 = exp1.to_i

      expenses.each_with_index do |exp2, idx2|
        next if idx == idx2
        @expense2 = exp2.to_i

        expenses.each_with_index do |exp3, idx3|
          next if idx == idx3
          next if idx2 == idx3
          @expense3 = exp3.to_i

          break if (expense1 + expense2 + expense3) == DESIRED_EXPENSE
        end
        break if (expense1 + expense2 + expense3) == DESIRED_EXPENSE
      end

      break if (expense1 + expense2 + expense3) == DESIRED_EXPENSE
    end

    expense1 * expense2 * expense3
  end

  def solve
    puts "Solution for 2 Expenses:"
    puts two_expenses
    puts "Solution for 3 Expenses:"
    puts three_expenses
  end

  private

  def reset_expenses
    @expense1 = @expense2 = nil
  end

end

@dir = Pathname.new(__FILE__).realpath.dirname

## Run
@file = @dir.join('../../shared/input_day01.txt')
reporter = ExpenseReporter.new(@file)
puts reporter.solve

File: 2020/python/01/main.py
import os

class ExpenseReporter(object):
    DESIRED_EXPENSE = 2020

    def __init__(self, filename):
        self.expenses = []
        for row in filename:
            self.expenses.append(int(row))
        filename.close()

    def two_expenses(self):
        entries = [0,0]
        current_expense = 0

        for idx, expense1 in enumerate(self.expenses):
            entries[0] = expense1

            for idx2, expense2 in enumerate(self.expenses):
                if idx == idx2:
                    continue

                entries[1] = expense2
                current_expense = self.__sum(entries)

                if current_expense == self.DESIRED_EXPENSE:
                    break

            if current_expense == self.DESIRED_EXPENSE:
                break

        return self.__multiply(entries)

    def three_expenses(self):
        entries = [0,0,0]
        current_expense = 0

        for idx, expense1 in enumerate(self.expenses):
            entries[0] = expense1

            for idx2, expense2 in enumerate(self.expenses):
                if idx == idx2:
                    continue
                entries[1] = expense2

                for idx3, expense3 in enumerate(self.expenses):
                    if (idx == idx3) or (idx2 == idx3):
                        continue
                    entries[2] = expense3
                    current_expense = self.__sum(entries)
                    if current_expense == self.DESIRED_EXPENSE:
                        break

                current_expense = self.__sum(entries)
                if current_expense == self.DESIRED_EXPENSE:
                    break

            current_expense = self.__sum(entries)
            if current_expense == self.DESIRED_EXPENSE:
                break

        return self.__multiply(entries)

    def execute(self):
        print("Solution for 2 Expenses:")
        print(self.two_expenses())
        print("Solution for 3 Expenses:")
        print(self.three_expenses())

    def __sum(self,entries):
        return reduce(lambda e1, e2: e1 + e2, entries)

    def __multiply(self,entries):
        return reduce(lambda e1, e2: e1 * e2, entries)

directory = os.getcwd()
path = open("{dir}/{file}".format(dir=directory, file="2020/shared/input_day01.txt"))
report = ExpenseReporter(path)

report.execute()
